### ISSUE TITLE
Collection onSnapshot callback firing in the wrong order (iOS), cannot set options for CollectionReference onSnapshot()

### ISSUE THREAD

**Issue by mitchellmcm27** [invertase/react-native-firebase#613]
### Issue

All of my onSnapshot callbacks for collections are getting called three times **when documents are deleted**, and in the wrong order. Logging the my collection onSnapshot calls reveals the following sequence of `metadata`, all fired within about 450ms.

1. {hasPendingWrites: false, fromCache: false}
2. {hasPendingWrites: false, fromCache: true}
3. {hasPendingWrites: false, fromCache: false}

The Firebase iOS SDK has tests to make sure the callback is fired in a certain sequence: https://github.com/firebase/firebase-ios-sdk/blob/a03dfd3c887f57c12cfa41478a3201846459afde/Firestore/Example/Tests/Integration/API/FIRDatabaseTests.m#L456

From there, it seems that the data `fromCache` should be delivered first.  This incorrect ordering causes items to blink in and out of my UI when I delete them, because the cached data is outdated and doesn't reflect the deleted item.

Also note that I don't particularly want to receive metadata changes, but I'm not able to pass the `options` argument to CollectionReference.onSnapshot().

A potential workaround would be to look at `querySnapshot.metadata.fromCache` and ignore the data from the cache, but I'm not sure what unintended side effects this would have.

### Environment

1. Application Target Platform: iOS

2. Development Operating System: macOS

3. Build Tools: Xcode 9.0.1

4. React Native version: 0.48.4

5. RNFirebase Version: 3.0.5

6. Firebase Module: firestore

**Comment by mitchellmcm27** [comment]
Edited to reflect that this only happens when documents are deleted from the collection. Adding a document to a collection seems to fire the onSnapshot callback correctly.

**Comment by chrisbianca** [comment]
@mitchellmcm27 I'll answer in two parts:

1) The ordering should be reaching the JS side of the bridge in the same order that it's received from the native SDK.  Are you able to put a break point in the iOS code for `RNFirebaseFirestoreCollectionReference`'s `onSnapshot` method and confirm that the data is coming out of the native SDK successfully?

2) You mention that you're not able to pass the options argument to `onSnapshot`.  You should be able to pass through an object as the first parameter like so:

`collectionRef.onSnapshot({ includeDocumentMetaDataChanges: false }, listener);`

**Comment by mitchellmcm27** [comment]
Hi @chrisbianca thanks for the reply.

1. Can do, will post the result below.

2. Maybe I'm missing something, but I am passing the options object as you showed, and I get the error below. It references CollectionRef.js and Query.js. It seems that the onSnapshot method of CollecitonRef does not take the options argument (see below), but I may be missing something. If it matters, I'm calling this inside of a `redux-thunk` action creator.

![image](https://user-images.githubusercontent.com/14998294/32976433-87019ada-cbe3-11e7-83fb-edb7f94af74c.png)

```js
// ...
firebase
  .firestore()
  .collection("users")
  .doc(uid)
  .collection("bars")
  .doc("default")
  .collection("bottles")
  .onSnapshot(
    { includeDocumentMetaDataChanges: false },
    querySnapshot => {
      console.log(querySnapshot.metadata);
      let bottles = {};
      querySnapshot.forEach(doc => {
        bottles[doc.id] = { ...doc.data() };
      });
      dispatch({ type: BOTTLES_SUCCESS, payload: bottles });
    },
    error => dispatch({ type: BOTTLES_FAILURE, payload: error })
  );
```

```js
// rnfirebase CollectionReference.js 

onSnapshot(onNext: () => any, onError?: () => any): () => void {
  return this._query.onSnapshot(onNext, onError);
}

```

```js
// rnfirebase Query.js

onSnapshot(
    optionsOrObserverOrOnNext: QueryListenOptions | Observer | (DocumentSnapshot) => void,
    observerOrOnNextOrOnError?: Observer | (DocumentSnapshot) => void | (Object) => void,
    onError?: (Object) => void,
) { 
// ...
}
```

**Comment by mitchellmcm27** [comment]
@chrisbianca Sorry for the delay, my Macbook has been unable to boot lately. I have to repair it. I do still plan to put the breakpoints in the iOS SDK and see what's going on.

**Comment by berkcoker** [comment]
I have the same problem actually. Only happens when a deletion is made. And I do get the same "First argument must be a function, observer or options error.

**Comment by berkcoker** [comment]
@chrisbianca I just did some debugging and the ordering is in fact correct. I wonder why this is happening and if there is a solution.

**Comment by berkcoker** [comment]
Contacted Firebase about it, and it looks like it is a known issue that they are trying to address. FYI

**Comment by chrisbianca** [comment]
@berkcoker thanks for confirming, I'm glad it's nothing our end!

**Comment by mitchellmcm27** [comment]
Thanks @berkcoker and @chrisbianca

**Comment by callagga** [comment]
I'm getting the "Error: Query.onSnapshot failed: First argument must be a function, observer or options." too.  A couple of questions:

a) So just to understand, we're still awaiting Firebase to correct before we can get granularity in the updates, so we can see both the initial offline response, and then when the app is online again if you want to pick up this 2nd update (so you could highlight which items in the list have been "synched" back to the server for example)?

b) I note on page https://rnfirebase.io/docs/v3.1.*/firestore/reference/CollectionReference#onSnapshot it mentioned the option name being "includeMetadataChanges", however is it "includeDocumentMetaDataChanges"?

```
firebase.firestore().collection('todos').onSnapshot(
      { includeDocumentMetaDataChanges: false },
      (querySnapshot) => {
        const items = querySnapshot.docs.map(doc => ({ 
          id: doc.id,
          fromCache: doc.metadata.fromCache,
          hasPendingWrites: doc.metadata.hasPendingWrites,
          title: doc.data().title,
        }));
        emit({ items, error: undefined });
      }

```

**Comment by chrisbianca** [comment]
Closing as a Firebase issue.
There is a fix for the options issue in our next release.

**Comment by sowdri** [comment]
@berkcoker I'm having the same issue as well. As you have mentioned that it's a firebase issue, is there a issue that I can track in firebase-github page? I couldn't find any issues filed at the moment.

**Comment by mitchellmcm27** [comment]
@sowdri I believe this is the issue https://github.com/firebase/firebase-ios-sdk/issues/1045

**Comment by mitchellmcm27** [comment]
Hopefully this has been fixed in Firestore 0.13.0 for iOS
https://github.com/firebase/firebase-ios-sdk/blob/master/Firestore/CHANGELOG.md
